Add Airflow sandbox docs to the data pipeline (#89)
moj-analytical-services/platform_user_guidance: 07-data_pipelines.Rmd

    e.g. See: https://github.com/moj-analytical-services/airflow-enforcement-data-engineering/blob/master/iam_policy.json
    for listing, reading and writing to S3 buckets.

## Test pipeline in your own Airflow sandbox

You can deploy your own sandboxed version of Airflow to test DAGs and docker images on.

### Start Airflow sandbox

> Click the `Deploy` button next to `airflow-sqlite` in the Analytical tools section of the control panel. This should take a few minutes to deploy. Then Click the `Open` button to open your Airflow sandbox in a new window.

Reviewer: ```suggestion
Click the `Deploy` button next to `airflow-sqlite` in the Analytical tools section of the control panel. This should take a few minutes to deploy. Then click the `Open` button to open your Airflow sandbox in a new window.
```

> Click the `Deploy` button next to `airflow-sqlite` in the Analytical tools section of the control panel. This should take a few minutes to deploy. Then click the `Open` button to open your Airflow sandbox in a new window.

Author: :+1:

> This will create an `airflow` directory in your home directory on the Analytical Platform. In this airflow directory there are three more directories `db`, `dags` and `logs`. The `dags` directory is where you will store your test dag files.

Reviewer: ```suggestion
This will create an `airflow` directory in your home directory on the Analytical Platform. In this airflow directory there are three more directories `db`, `dags` and `logs`. The `dags` directory is where you will store your test DAG files.
```

> This will create an `airflow` directory in your home directory on the Analytical Platform. In this airflow directory there are three more directories `db`, `dags` and `logs`. The `dags` directory is where you will store your test DAG files.

Author: :+1:

> This will create an `airflow` directory in your home directory on the Analytical Platform. In this airflow directory there are three more directories `db`, `dags` and `logs`. The `dags` directory is where you will store your test DAG files.

Reviewer 2: Think there needs a section (maybe here) that just quickly outlines the differences between this dag and the production one:
- You must use your `iam_role` (I know that it's written about later but might be worth pointing out here)
- Tasks can only run one at a time

### Create a test DAG

> Using JupiterLab on the Analytical Platform create a python file in the `airflow/dags` directory in your home directory in the Analytical Platform (eg. `test_dag.py`). Airflow will automatically scan this directory for dag files every three minutes.

Reviewer: ```suggestion
Using JupiterLab on the Analytical Platform create a python file in the `airflow/dags` directory in your home directory in the Analytical Platform (eg. `test_dag.py`). Airflow will automatically scan this directory for DAG files every three minutes.
```

> Using JupyterLab on the Analytical Platform create a python file in the `airflow/dags` directory in your home directory in the Analytical Platform (eg. `test_dag.py`). Airflow will automatically scan this directory for DAG files every three minutes.

Author: :+1:

> See above for creating a Docker image to run the dag if you are using `KubernetesPodOperator`. Concourse will automatically build images of PR's to github airflow repos and will be tagged with the name of the PR.

Reviewer: ```suggestion
See above for creating a Docker image to run the DAG if you are using `KubernetesPodOperator`. Concourse will automatically build images of PR's to github airflow repos and will be tagged with the name of the PR.
```

> See above for creating a Docker image to run the DAG if you are using `KubernetesPodOperator`. Concourse will automatically build images of PR's to GitHub Airflow repos and will be tagged with the name of the PR.

Author: 👍

> See above for creating a Docker image to run the DAG if you are using `KubernetesPodOperator`. Concourse will automatically build images of PR's to GitHub Airflow repos and will be tagged with the name of the PR.

Reviewer 2: Might be worth adding an example for this e.g.

> If your pull request is named dev then concourse will build an image named `my_repositiry:dev`.

Might be over explaining though...

Example DAG file contents:

    from datetime import datetime

    from airflow.contrib.operators.kubernetes_pod_operator import KubernetesPodOperator
    from airflow.utils.log.logging_mixin import LoggingMixin
    from airflow.models import DAG

    log = LoggingMixin().log

    args = {"owner": "airflow", "start_date": datetime(2019, 3, 18)}

    dag = DAG(
        dag_id="example_kubernetes_operator_assume_role",
        default_args=args,
        schedule_interval=None,
    )

    k = KubernetesPodOperator(
        namespace="user-###YOUR_GITHUB_USERNAME###",
        image="governmentpaas/awscli:latest",
        cmds=["aws", "--debug"],
        arguments=["s3api", "list-objects", "--bucket", "###YOUR_TEST_BUCKET###"],
        env_vars={
            "AWS_METADATA_SERVICE_TIMEOUT": "60",
            "AWS_METADATA_SERVICE_NUM_ATTEMPTS": "5",
        },
        labels={"foo": "bar"},
        name="airflow-test-pod",
        in_cluster=True,
        task_id="task",
        get_logs=True,
        dag=dag,
        is_delete_operator_pod=True,
        annotations={"iam.amazonaws.com/role": "alpha_user_###YOUR_GITHUB_USERNAME###"},
    )

> You will have to set the dag namespace to your namespace and the aws role to your role, see `###YOUR_GITHUB_USERNAME###` in the above example code.

Reviewer: ```suggestion
You will have to set the DAG namespace to your namespace and the AWS role to your role, see `###YOUR_GITHUB_USERNAME###` in the above example code.
```

> You will have to set the DAG namespace to your namespace and the AWS role to your role, see `###YOUR_GITHUB_USERNAME###` in the above example code.

Author: :+1:

> You will have to set the DAG namespace to your namespace and the AWS role to your role, see `###YOUR_GITHUB_USERNAME###` in the above example code.

Author: I am not sure if I can provide an internal link in to a code sample.

> You will also need to update the image with your dag image from AWS ECR and add any environment variables etc. as you would with a normal DAG.

Reviewer: ```suggestion
You will also need to update the image with your DAG image from AWS ECR and add any environment variables etc. as you would with a normal DAG.
```

> You will also need to update the image with your DAG image from AWS ECR and add any environment variables etc. as you would with a normal DAG.

Author: :+1:

> ### IMPORTANT

Reviewer 3: Could you change this to something like `Important considerations`?

Author: Yeah sure. Do you think in all caps? It is very important that people read and understand them.

> This sandbox Airflow is meant for testing. You should not use it to run regular tasks and should not be relied on for running production pipelines. It is likely that the sandbox will be idled when it is not being used.

Author: :+1:

> Because you are running this DAG as your own `alpha_user_...`, the dag will have the access that your user has. It does not run with the `iam_policy.json` that you may have created in the task github repo. This means that you might need to either grant yourself access to any buckets or copy data in to a bucket you have created in the Warehouse data section of the Analytical Platform.

Reviewer: ```suggestion
Because you are running this DAG as your own `alpha_user_...`, the DAG will have the access that your user has. It does not run with the `iam_policy.json` that you may have created in the task GitHub repo. This means that you might need to either grant yourself access to any buckets or copy data in to a bucket you have created in the Warehouse data section of the Analytical Platform.
```

> Because you are running this DAG as your own `alpha_user_...`, the DAG will have the access that your user has. It does not run with the `iam_policy.json` that you may have created in the task GitHub repo. This means that you might need to either grant yourself access to any buckets or copy data in to a bucket you have created in the Warehouse data section of the Analytical Platform.

Reviewer 3: ```suggestion
Because DAGs are run as your own `alpha_user_...`, they will have the same access permissions. DAGs are not run using an `iam_policy.json`, even if one exists in the task GitHub repository. This means that you may need to either obtain access to any required buckets or copy data into a bucket that you already have access to.
```

Author: :+1:

Reviewer: I reckon this might usefully have another sentence on @aldo's point i.e. to warn that your role can do potentially do bad things, with likely access to buckets across all your projects.

You can manually start ("trigger") the pipeline with this button:

![](images/pipelines/trigger.png)

## Test pipeline in your own Airflow sandbox

You can deploy your own sandboxed version of Airflow to test DAGs and docker images on.

### Start Airflow sandbox

Click the `Deploy` button next to `airflow-sqlite` in the Analytical tools section of the control panel. This should take a few minutes to deploy. Then Click the `Open` button to open your Airflow sandbox in a new window.

This will create an `airflow` directory in your home directory on the Analytical Platform. In this airflow directory there are three more directories `db`, `dags` and `logs`. The `dags` directory is where you will store your test dag files.

### Create a test DAG

> Using Jupiter lab on the Analytical Platform create a python file in the `airflow/dags` directory in your home directory in the Analytical Platform (eg. `test_dag.py`). Airflow will automatically scan this directory for dag files every three minutes.

Reviewer: ```suggestion
Use JupyterLab on the Analytical Platform to create a python file in the `airflow/dags` directory in your home directory in the Analytical Platform (eg `test_dag.py`). Airflow will automatically scan this directory for DAG files every three minutes.
```

    args = {"owner": "airflow", "start_date": datetime(2019, 3, 18)}

    dag = DAG(
        dag_id="example_kubernetes_operator_assume_role",
        default_args=args,
        schedule_interval=None,
    )

    k = KubernetesPodOperator(
        namespace="user-###YOUR_GITHUB_USERNAME###",
        image="governmentpaas/awscli:latest",
        cmds=["aws", "--debug"],
        arguments=["s3api", "list-objects", "--bucket", "###YOUR_TEST_BUCKET###"],
        env_vars={
            "AWS_METADATA_SERVICE_TIMEOUT": "60",
            "AWS_METADATA_SERVICE_NUM_ATTEMPTS": "5",
        },
        labels={"foo": "bar"},
        name="airflow-test-pod",
        in_cluster=True,
        task_id="task",
        get_logs=True,
        dag=dag,
        is_delete_operator_pod=True,
        annotations={"iam.amazonaws.com/role": "alpha_user_###YOUR_GITHUB_USERNAME###"},
    )

You will have to set the dag namespace to your namespace and the aws role to your role, see `###YOUR_GITHUB_USERNAME###` in the above example code.

You will also need to update the image with your dag image from AWS ECR and add any environment variables etc. as you would with a normal DAG.

### IMPORTANT

> This sandbox Airflow is meant for testing. You should not use it to run regular tasks and s should not be relied on for running production pipelines. It is likely that the sandbox will be idled when it is not being used.

Reviewer: ```suggestion
This sandbox Airflow is meant for testing. You should not use it to run regular tasks and should not be relied on for running production pipelines. It is likely that the sandbox will be idled when it is not being used.
```

You can manually start ("trigger") the pipeline with this button:

![](images/pipelines/trigger.png)

## Test pipeline in your own Airflow sandbox

> You can deploy your own sandboxed version of Airflow to test Dags and docker images on.

Reviewer 4: Hope it doesn't sound like I'm splitting hairs but we should try and keep the case of DAGs as "DAGs" as it is an acronym (directed acyclic graph)

Reviewer 4: Also, maybe write it as DAG (directed acyclic graph) the first time for the people that don't know what DAG stands for.

Author: :+1:

### Start Airflow sandbox

Click the `Deploy` button next to `airflow-sqlite` in the Analytical tools section of the control panel. This should take a few minutes to deploy. Then Click the `Open` button to open your Airflow sandbox in a new window.

> This will creat an `airflow` directory in your home directory on the Analytical Platform. In this airflow directory there are three more directories `db`, `dags` and `logs`. The `dags` directory is where you will store your test dag files.

Reviewer 4: creat => create
dag => DAG

> ### Create a test Dag

Reviewer 4: Dag => DAG 🤣

Using Jupiter lab on the Analytical Platform create a python file in the `airflow/dags` directory in your home directory in the Analytical Platform (eg. `test_dag.py`). Airflow will automatically scan this directory for dag files every three minutes.

See above for creating a Docker image to run the dag if you are using `KubernetesPodOperator`. Concourse will automatically build images of PR's to github airflow repos and will be tagged with the name of the PR.

Example Dag file contents:

    from airflow.utils.dates import days_ago
    from airflow.utils.log.logging_mixin import LoggingMixin
    from airflow.models import DAG

    log = LoggingMixin().log

>     try:

Reviewer 4: Do you need to import `KubernetesPodOperator` in a `try` block? In theory it should be there so we can simplify the example (unless I'm missing something of course)

        from airflow.contrib.operators.kubernetes_pod_operator import KubernetesPodOperator

>         args = {"owner": "airflow", "start_date": days_ago(2)}

Reviewer 4: @isichei what's the latest guideline on dynamic `start_date` in DAGs 🤣 ?

            default_args=args,
            schedule_interval=None,
        )

        k = KubernetesPodOperator(
            namespace="user-###YOUR_GITHUB_USERNAME###",
            image="governmentpaas/awscli:latest",
            cmds=["aws", "--debug"],
            arguments=["s3api", "list-objects", "--bucket", "###YOUR_TEST_BUCKET###"],
            env_vars={
                "AWS_METADATA_SERVICE_TIMEOUT": "60",
                "AWS_METADATA_SERVICE_NUM_ATTEMPTS": "5",
            },
            labels={"foo": "bar"},
            name="airflow-test-pod",
            in_cluster=True,
            task_id="task",
            get_logs=True,
            dag=dag,
            annotations={"iam.amazonaws.com/role": "alpha_user_###YOUR_GITHUB_USERNAME###"},
        )

    except ImportError as e:
        log.warn("Could not import KubernetesPodOperator: " + str(e))

You will have to set the dag namespace to your namespace and the aws role to your role, see `###YOUR_GITHUB_USERNAME###` in the above example code.

You will also need to update the image with your dag image from AWS ECR and add any environment variables etc. as you would with a normal Dag.

> ### Caveats

Reviewer 4: I'd probably change this to **IMPORTANT** as we want to be very sure users will not start using this to run jobs as an alternative to the main Airflow.

This sandbox Airflow is meant for testing. You should not use it to run regular tasks and s should not be relied on for running production pipelines. It is likely that the sandbox will be idled when it is not being used.

> Because you are running this Dag as your own `alpha_user_...`, the dag will only have the access that your user has. It does not run with the `iam_policy.json` that you may have created in the task github repo. This means that you might need to either grant yourself access to any buckets or copy data in to a bucket you have created in the Warehouse data section of the Analytical Platform.

Reviewer 4: "dag" => "DAG" (sorry :P)

But most important, I'd remove the "only" as it sounds like the DAG will run with not much privileges while that's actually the opposite. My English is not great but we should probably be very very very explicit about the fact that the code will run as their role and potentially do bad things as themselves.

This is my main worry, that because this is a "sandbox"/test airflow they'll not realise that they can do a lot of damage as the code runs like themselves. Don't know if you know what I mean.

            task_id="task",
            get_logs=True,
            dag=dag,

>             is_delete_operator_pod=True,

Reviewer 4: Simplest solution 👍
